demos/sample_azure_iot_adu/sample_azure_iot_adu.c

```diff
@@ -918,33 +888,72 @@ static void prvAzureDemoTask( void * pvParameters )
+                                                                 sampleazureiotPROCESS_LOOP_TIMEOUT_MS );
+                        configASSERT( xResult == eAzureIoTSuccess );
+
+                        /* In prvDownloadUpdateImageIntoFlash, we make a call to the _ProcessLoop() function */
+                        /* which could bring in a new or cancelled update. */
+                        /* Check xProcessUpdateRequest and the action again in case a new version came in that was invalid. */
+                        if( xProcessUpdateRequest && ( xAzureIoTAduUpdateRequest.xWorkflow.xAction == eAzureIoTADUActionApplyDownload ) )
+                        {
+                            xResult = prvEnableImageAndResetDevice();
+                            configASSERT( xResult == eAzureIoTSuccess );
+
+                            xResult = prvSpoofNewVersion();
+                            configASSERT( xResult == eAzureIoTSuccess );
+                        }
+                        else
+                        {
+                            xResult = AzureIoTADUClient_SendAgentState( &xAzureIoTADUClient,
+                                                                        &xAzureIoTHubClient,
+                                                                        &xADUDeviceProperties,
+                                                                        &xAzureIoTAduUpdateRequest,
+                                                                        eAzureIoTADUAgentStateIdle,
+                                                                        NULL,
+                                                                        ucScratchBuffer,
+                                                                        sizeof( ucScratchBuffer ),
+                                                                        NULL );
+
+                            xProcessUpdateRequest = false;
+                        }
+                    }
+                    else
+                    {
+                        LogInfo( ( "Unknown action received: %i", xAzureIoTAduUpdateRequest.xWorkflow.xAction ) );
+                    }
                 }
-                else
-                {
-                    LogInfo( ( "Unknown action received: %i", xAzureIoTAduUpdateRequest.xWorkflow.xAction ) );
-                }
-            }
 
-            /* Leave Connection Idle for some time. */
-            LogInfo( ( "Keeping Connection Idle..." ) );
-            vTaskDelay( sampleazureiotDELAY_BETWEEN_PUBLISHES_TICKS );
-        }
+                /* Leave Connection Idle for some time. */
+                LogInfo( ( "Keeping Connection Idle..." ) );
+                vTaskDelay( sampleazureiotDELAY_BETWEEN_PUBLISHES_TICKS );
+            }
 
-        xResult = AzureIoTHubClient_UnsubscribeProperties( &xAzureIoTHubClient );
-        configASSERT( xResult == eAzureIoTSuccess );
+            if( xAzureSample_IsConnectedToInternet() )
```

Comment:
Before, we couldn't get to this point because of the for (; ; ) - this will change the behavior of the sample now that this code is reachable

Reply:
Correct. In reality this could still be unreachable, since to get here there must NOT be internet connection. And then if the check again confirms there is still no internet connection, the graceful disconnection commands will not be run. To improve this we would need to have further refactoring.